Lint: new files_unchanged test to check that files are not modified from the template

ewels opened the pull request:
There are quite a lot of files in the nf-core template that should basically be left untouched. As a result we don't really check them much and when stuff is broken it's quite hard to debug (often due to a template update merge gone wrong).

This is a simple new linting test that takes a list of files and checks their contents against a fresh template output. The majority should be identical, there is also a list for those that can have additions (currently just `.gitignore`).

This test is currently very DSL1 centric, just because that's what is in the template right now. It will be trivial to update to DSL2 when the time comes though.

Note that I have removed the `licence` lint test as this is now superseded by tests that check that the file is there and is unmodified.

## New feature: `--fix`

I've also added a new flag for `nf-core lint` called `--fix`. We know what these files _should_ look like so it should be simple to overwrite what's currently in the pipeline with what should be there. This will make getting that little green tick ever so much simpler. We should be able to extend this functionality to other lint tests too (eg. version consistency checks etc).

## PR checklist

 - [x] This comment contains a description of changes (with reason)
 - [x] `CHANGELOG.md` is updated
 - [ ] If you've fixed a bug or added code that should be tested, add tests!
 - [ ] Documentation in `docs` is updated

ewels commented:
I wonder if we can remove a bunch of other tests with this in place - eg. all of the various tests for the AWS workflows. Need to think about how much we expect people to change these / care about that. Some also need customisation (eg. the full size tests workflow) so can't be ditched. But I think that the branch tests can probably go? 'cc @ggabernet 

Also note that we could probably implement `--fix` for a bunch of these fairly easily - just overwrite the parsed yaml with what we expect and then save that.

ggabernet commented:
Nice, that will be a great addition to the `lint` functionality. I'm reviewing now, I have some comments / questions though, e.g. if some of the files such as GitHub actions workflows cannot be edited then we should remove the `TODO` strings from them.
Or would it be desirable that the CI action is not edited? Then people cannot add extra tests there?
I'll comment case by case on the review

ggabernet commented on nf_core/lint/files_unchanged.py:
```diff
@@ -0,0 +1,202 @@
+        '.github/ISSUE_TEMPLATE/bug_report.md',
+        '.github/ISSUE_TEMPLATE/config.yml',
+        '.github/ISSUE_TEMPLATE/feature_request.md',
+        '.github/workflows/awsfulltest.yml',
```
I agree but then the `TODO` string should be removed from here on the template, and `-profile test` changed to `-profile test_full`

ewels replied:
I guess that the only down side of this is if we want people to be able to run _multiple_ AWS full size tests (eg. the `rnaseq` pipeline with different aligners).

Though actually if we move this to the `files_partial` list instead then it should be fine as people could add additional workflow steps below it.. Ok I'll play with the workflow files a little then :+1:

ewels replied:
Yeah, same as you say below in https://github.com/nf-core/tools/pull/846/files#r573026542 I realise..

Problem with this is that people can't edit the `strategy` `matrix`.

I wonder if we're better ditching these files from this test but rewriting the existing tests that we have for them to be a bit simpler - just checking for equivalence in the YAML for the majority of keys.

ewels replied:
Removed from this test 👍🏻

ggabernet replied:
ok yeah that sounds good

ggabernet commented on nf_core/lint/files_unchanged.py:
```diff
@@ -0,0 +1,202 @@
+        '.github/ISSUE_TEMPLATE/config.yml',
+        '.github/ISSUE_TEMPLATE/feature_request.md',
+        '.github/workflows/awsfulltest.yml',
+        '.github/workflows/awstest.yml',
```
`TODO` string should be removed from the template then, if it can't be edited

ewels replied:
File removed from test 👍🏻

ewels commented:
Ah the AWS test actions should be removed from this check, yes - that's my mistake. As you say this should only be for files where we don't want people to edit the file contents.

ggabernet commented on nf_core/lint/files_unchanged.py:
```diff
@@ -0,0 +1,202 @@
+        [os.path.join(".github", "workflows", "awsfulltest.yml")],
+        [os.path.join(".github", "workflows", "awstest.yml")],
+        [os.path.join(".github", "workflows", "branch.yml")],
+        [os.path.join(".github", "workflows", "ci.yml")],
```
wouldn't it be desirable to add further tests in the `ci.yml` so it would be a file that must contain the template content but can contain other stuff

ewels replied:
File removed from this test 👍🏻

ggabernet commented on nf_core/lint/files_unchanged.py:
```diff
@@ -0,0 +1,202 @@
+        [os.path.join(".github", "workflows", "awsfulltest.yml")],
+        [os.path.join(".github", "workflows", "awstest.yml")],
+        [os.path.join(".github", "workflows", "branch.yml")],
+        [os.path.join(".github", "workflows", "ci.yml")],
```
```suggestion
```

ggabernet commented on nf_core/lint/files_unchanged.py:
```diff
@@ -0,0 +1,202 @@
+    ]
+    files_partial = [
+        [".gitignore", "foo"],
+        [os.path.join("assets", "multiqc_config.yaml")],
```
```suggestion
        [os.path.join("assets", "multiqc_config.yaml")],
        [os.path.join(".github", "workflows", "ci.yml")],
        [os.path.join(".github", "workflows", "push_dockerhub_dev.yml")],
        [os.path.join(".github", "workflows", "push_dockerhub_release.yml")],
```

ggabernet commented on nf_core/lint/files_unchanged.py:
```diff
@@ -0,0 +1,202 @@
+        [os.path.join(".github", "workflows", "ci.yml")],
+        [os.path.join(".github", "workflows", "linting.yml")],
+        [os.path.join(".github", "workflows", "linting_comment.yml")],
+        [os.path.join(".github", "workflows", "push_dockerhub_dev.yml")],
+        [os.path.join(".github", "workflows", "push_dockerhub_release.yml")],
```
some pipelines have additional containers (e.g. mag pipeline) because there are depenency issues, so maybe this would be better added to files that should contain the template but can contain extra steps?

ewels replied:
Done 👍🏻

ggabernet commented on nf_core/lint/files_unchanged.py:
```diff
@@ -0,0 +1,202 @@
+        [os.path.join(".github", "workflows", "ci.yml")],
+        [os.path.join(".github", "workflows", "linting.yml")],
+        [os.path.join(".github", "workflows", "linting_comment.yml")],
+        [os.path.join(".github", "workflows", "push_dockerhub_dev.yml")],
+        [os.path.join(".github", "workflows", "push_dockerhub_release.yml")],
```
```suggestion
```

ggabernet commented:
> Ah the AWS test actions should be removed from this check, yes - that's my mistake. As you say this should only be for files where we don't want people to edit the file contents.

Actually if we just make the 2 edits I suggested for the actions workflow `awsfulltest.yml` and `awstest.yml` then they could be included in this test function and the other lint tests can be removed. The user the only needs to edit the `test_full.config` in any case. The lint tests for the `branch.yml` could be then also be removed as you suggested.

ewels commented:
See inline comments also, but the problem with just adding to the end of these files is that then users can't use `strategy` `matrix` and other aspects of the workflow for job parallelisation. I think that if we suspect that users will be editing files significantly then they shouldn't be included in this test and we can do a nicer job with the test that parses the YAML.

ewels commented:
Ok, deleted some more lint tests that are no longer needed with this file comparison.

I also added some new `--fix` functionality: it can now automatically bump all software versions to the latest available in `environment.yml`.

One more `--fix` addition that would be nice is to add in missing files in the `files_exist` test and delete any that are not supposed to be there. This needs a little refactoring though as we need the pipeline config to generate the template and at the moment that's not yet done when we check that the files exist. It would also be good to generate a template pipeline only once instead of multiple times. Bed now but will come back to that.

I'm feeling pretty good about this! I think that this `--fix` functionality could actually make the `TEMPLATE` merges much easier! By doing all the basic stuff here first the diff should be a lot smaller making the merge quite a bit easier.

MaxUlysse commented:
Run `nf-core . lint --fix` on the current sarek `dev`.
It is a bit too efficient I think.
cf: https://github.com/MaxUlysse/nf-core_sarek/commit/13acc6539e86813abfe0a19d13d986f97c2a993a

but nothing that is not easily fixable on my side.
I added a `title` property to the `bug_report.md` and `feature_request.md` (which I can also add to the template actually).
I added the `assignees` property which I'd really like to keep, makes it easier not to miss any issue.
Already made #847 to fix the `PULL_REQUEST_TEMPLATE.md ` issue.
The `docs/images/nf-core-sarek_logo.png` should not be changed, I do like the Sarek logo, and would very like to keep it, and I'm pretty sure @jfy133 will feel the same for eager.
And the update of every tool in the `environment.yml` file could be a nice feature, but I'd like that to be in a separate command.

ggabernet commented:
I was also going to test the `--fix` with Bcellmagic, but got this when running `nf-core lint .`

```
$ nf-core lint .

                                          ,--./,-.
          ___     __   __   __   ___     /,-._.--~\
    |\ | |__  __ /  ` /  \ |__) |__         }  {
    | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                          `._,._,'

    nf-core/tools version 1.13.dev0



╭───────────────────────────────────────────────── Traceback (most recent call last) ─────────────────────────────────────────────────╮
│ /Users/gisela/Projects/Pipelines/nf-core-tools/nf_core/utils.py:204 in fetch_wf_config                                              │
│                                                                                                                                     │
│   201 │   # Call `nextflow config` and pipe stderr to /dev/null                                                                     │
│   202 │   try:                                                                                                                      │
│   203 │   │   with open(os.devnull, "w") as devnull:                                                                                │
│ ❱ 204 │   │   │   nfconfig_raw = subprocess.check_output(["nextflow", "config", "-flat",                                            │
│       wf_path], stderr=devnull)                                                                                                     │
│   205 │   except OSError as e:                                                                                                      │
│   206 │   │   if e.errno == errno.ENOENT:                                                                                           │
│   207 │   │   │   raise AssertionError("It looks like Nextflow is not installed. It is                                              │
│       required for most nf-core functions.")                                                                                        │
│                                                                                                                                     │
│ /Users/gisela/miniconda3/envs/nfcore-tools-test/lib/python3.9/subprocess.py:420 in check_output                                     │
│                                                                                                                                     │
│    417 │   │   # empty string. That is maintained here for backwards compatibility.                                                 │
│    418 │   │   kwargs['input'] = '' if kwargs.get('universal_newlines', False) else b''                                             │
│    419 │                                                                                                                            │
│ ❱  420 │   return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,                                                         │
│    421 │   │   │      **kwargs).stdout                                                                                              │
│                                                                                                                                     │
│ /Users/gisela/miniconda3/envs/nfcore-tools-test/lib/python3.9/subprocess.py:524 in run                                              │
│                                                                                                                                     │
│    521 │   │   │   raise                                                                                                            │
│    522 │   │   retcode = process.poll()                                                                                             │
│    523 │   │   if check and retcode:                                                                                                │
│ ❱  524 │   │   │   raise CalledProcessError(retcode, process.args,                                                                  │
│    525 │   │   │   │   │   │   │   │   │    output=stdout, stderr=stderr)                                                           │
│    526 │   return CompletedProcess(process.args, retcode, stdout, stderr)                                                           │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
CalledProcessError: Command '['nextflow', 'config', '-flat', '.']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

╭───────────────────────────────────────────────── Traceback (most recent call last) ─────────────────────────────────────────────────╮
│ /Users/gisela/miniconda3/envs/nfcore-tools-test/bin/nf-core:33 in <module>                                                          │
│                                                                                                                                     │
│ no lexer for filename 'nf-core' found                                                                                               │
│                                                                                                                                     │
│ /Users/gisela/Projects/Pipelines/nf-core-tools/nf_core/__main__.py:56 in run_nf_core                                                │
│                                                                                                                                     │
│    53 │   stderr.print("\n\n")                                                                                                      │
│    54 │                                                                                                                             │
│    55 │   # Lanch the click cli                                                                                                     │
│ ❱  56 │   nf_core_cli()                                                                                                             │
│    57                                                                                                                               │
│    58                                                                                                                               │
│    59 # Customise the order of subcommands for --help                                                                               │
│                                                                                                                                     │
│ /Users/gisela/miniconda3/envs/nfcore-tools-test/lib/python3.9/site-packages/click/core.py:829 in __call__                           │
│                                                                                                                                     │
│    826 │                                                                                                                            │
│    827 │   def __call__(self, *args, **kwargs):                                                                                     │
│    828 │   │   """Alias for :meth:`main`."""                                                                                        │
│ ❱  829 │   │   return self.main(*args, **kwargs)                                                                                    │
│    830                                                                                                                              │
│    831                                                                                                                              │
│    832 class Command(BaseCommand):                                                                                                  │
│                                                                                                                                     │
│ /Users/gisela/miniconda3/envs/nfcore-tools-test/lib/python3.9/site-packages/click/core.py:782 in main                               │
│                                                                                                                                     │
│    779 │   │   try:                                                                                                                 │
│    780 │   │   │   try:                                                                                                             │
│    781 │   │   │   │   with self.make_context(prog_name, args, **extra) as ctx:                                                     │
│ ❱  782 │   │   │   │   │   rv = self.invoke(ctx)                                                                                    │
│    783 │   │   │   │   │   if not standalone_mode:                                                                                  │
│    784 │   │   │   │   │   │   return rv                                                                                            │
│    785 │   │   │   │   │   # it's not safe to `ctx.exit(rv)` here!                                                                  │
│                                                                                                                                     │
│ /Users/gisela/miniconda3/envs/nfcore-tools-test/lib/python3.9/site-packages/click/core.py:1259 in invoke                            │
│                                                                                                                                     │
│   1256 │   │   │   │   Command.invoke(self, ctx)                                                                                    │
│   1257 │   │   │   │   sub_ctx = cmd.make_context(cmd_name, args, parent=ctx)                                                       │
│   1258 │   │   │   │   with sub_ctx:                                                                                                │
│ ❱ 1259 │   │   │   │   │   return _process_result(sub_ctx.command.invoke(sub_ctx))                                                  │
│   1260 │   │                                                                                                                        │
│   1261 │   │   # In chain mode we create the contexts step by step, but after the                                                   │
│   1262 │   │   # base command has been invoked.  Because at that point we do not                                                    │
│                                                                                                                                     │
│ /Users/gisela/miniconda3/envs/nfcore-tools-test/lib/python3.9/site-packages/click/core.py:1066 in invoke                            │
│                                                                                                                                     │
│   1063 │   │   """                                                                                                                  │
│   1064 │   │   _maybe_show_deprecated_notice(self)                                                                                  │
│   1065 │   │   if self.callback is not None:                                                                                        │
│ ❱ 1066 │   │   │   return ctx.invoke(self.callback, **ctx.params)                                                                   │
│   1067                                                                                                                              │
│   1068                                                                                                                              │
│   1069 class MultiCommand(Command):                                                                                                 │
│                                                                                                                                     │
│ /Users/gisela/miniconda3/envs/nfcore-tools-test/lib/python3.9/site-packages/click/core.py:610 in invoke                             │
│                                                                                                                                     │
│    607 │   │   args = args[2:]                                                                                                      │
│    608 │   │   with augment_usage_errors(self):                                                                                     │
│    609 │   │   │   with ctx:                                                                                                        │
│ ❱  610 │   │   │   │   return callback(*args, **kwargs)                                                                             │
│    611 │                                                                                                                            │
│    612 │   def forward(*args, **kwargs):  # noqa: B902                                                                              │
│    613 │   │   """Similar to :meth:`invoke` but fills in default keyword                                                            │
│                                                                                                                                     │
│ /Users/gisela/Projects/Pipelines/nf-core-tools/nf_core/__main__.py:318 in lint                                                      │
│                                                                                                                                     │
│   315 │   """                                                                                                                       │
│   316 │                                                                                                                             │
│   317 │   # Run the lint tests!                                                                                                     │
│ ❱ 318 │   lint_obj = nf_core.lint.run_linting(pipeline_dir, release, fix, show_passed,                                              │
│       markdown, json)                                                                                                               │
│   319 │   if len(lint_obj.failed) > 0:                                                                                              │
│   320 │   │   sys.exit(1)                                                                                                           │
│                                                                                                                                     │
│ /Users/gisela/Projects/Pipelines/nf-core-tools/nf_core/lint/__init__.py:47 in run_linting                                           │
│                                                                                                                                     │
│    44 │                                                                                                                             │
│    45 │   # Load the various pipeline configs                                                                                       │
│    46 │   lint_obj._load_lint_config()                                                                                              │
│ ❱  47 │   lint_obj._load_pipeline_config()                                                                                          │
│    48 │   lint_obj._load_conda_environment()                                                                                        │
│    49 │   lint_obj._list_files()                                                                                                    │
│                                                                                                                                     │
│ /Users/gisela/Projects/Pipelines/nf-core-tools/nf_core/utils.py:131 in _load_pipeline_config                                        │
│                                                                                                                                     │
│   128 │   │                                                                                                                         │
│   129 │   │   Once loaded, set a few convienence reference class attributes                                                         │
│   130 │   │   """                                                                                                                   │
│ ❱ 131 │   │   self.nf_config = fetch_wf_config(self.wf_path)                                                                        │
│   132 │   │                                                                                                                         │
│   133 │   │   self.pipeline_name = self.nf_config.get("manifest.name",                                                              │
│       "").strip("'").replace("nf-core/", "")                                                                                        │
│                                                                                                                                     │
│ /Users/gisela/Projects/Pipelines/nf-core-tools/nf_core/utils.py:209 in fetch_wf_config                                              │
│                                                                                                                                     │
│   206 │   │   if e.errno == errno.ENOENT:                                                                                           │
│   207 │   │   │   raise AssertionError("It looks like Nextflow is not installed. It is                                              │
│       required for most nf-core functions.")                                                                                        │
│   208 │   except subprocess.CalledProcessError as e:                                                                                │
│ ❱ 209 │   │   raise AssertionError("`nextflow config` returned non-zero error code: %s,\n                                           │
│       %s", e.returncode, e.output)                                                                                                  │
│   210 │   else:                                                                                                                     │
│   211 │   │   for l in nfconfig_raw.splitlines():                                                                                   │
│   212 │   │   │   ul = l.decode("utf-8")                                                                                            │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
AssertionError: ('`nextflow config` returned non-zero error code: %s,\n   %s', 1, b'')
```

ewels commented:
@ggabernet - tempted to say "works as expected" 😆  but yeah, I guess it could be handled with a nicer error message. I'll look into adding that now.

@MaxUlysse - see below:

> And the update of every tool in the `environment.yml` file could be a nice feature, but I'd like that to be in a separate command.

I'm thinking that `--fix` should always be run with an argument to specify which lint test should be fixed. Then if lint tests _can_ be fixed the log can show a full command with all of them specified for easy copying + pasting. I think that's a good compromise, more explicit about what's going to happen and more control, but still easy.

> The `docs/images/nf-core-sarek_logo.png` should not be changed, I do like the Sarek logo, and would very like to keep it, and I'm pretty sure @jfy133 will feel the same for eager.

That's fine, but as it only really affects 2 (ok max a handful) of pipelines out of the 40 odd that we have, I think you guys can just do a `git checkout docs/images/nf-core-sarek_logo.png` to restore the clobbered file. I feel like it does the right thing for most people. I thought it was fairly important to have as we have changed the logo script recently so the images will now be different for all pipelines (white background).

> I added the `assignees` property which I'd really like to keep, makes it easier not to miss any issue.

This can be done in lots of different ways, so I'd prefer to keep the check and you can just use `CODEOWNERS` or an app or something. I think that should work as well, right?

ewels commented:
Ah wait, is the logo triggering a lint failure? Yeah I guess I could downgrade it to a warning 🤔 Or maybe better still you can just configure the linting to skip the check for that file! 🎉 

Make a root-level `.nf-core-lint.yml` file:

```
files_unchanged:
  - docs/images/nf-core-sarek_logo.png
```

Curious to see if this works!

ewels commented:
Screaming children but have done the nicer log message @ggabernet - if your nextflow is still broken would be great if you could test 😁

ggabernet commented:
Yeah I have a very reproducible way to break it, specifying JAVA_HOME to a version >11 😄 

Great, now it shows:
```
nf-core lint .

                                          ,--./,-.
          ___     __   __   __   ___     /,-._.--~\
    |\ | |__  __ /  ` /  \ |__) |__         }  {
    | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                          `._,._,'

    nf-core/tools version 1.13.dev0



CRITICAL ('`nextflow config` returned non-zero error code: %s,\n       __main__.py:323
         %s', 1, b'')
```

ggabernet commented:
Now I could also test your `--fix` feature. Amazing, that worked nicely for me and it will really help with template updates!

I think I agree with Maxime that having the versions of the tools updated could break the container due to dependency conflicts, but if this will be handled with specific arguments it will be still nice to have.

MaxUlysse commented:
> Ah wait, is the logo triggering a lint failure? Yeah I guess I could downgrade it to a warning thinking Or maybe better still you can just configure the linting to skip the check for that file! tada
> 
> Make a root-level `.nf-core-lint.yml` file:
> 
> ```
> files_unchanged:
>   - docs/images/nf-core-sarek_logo.png
> ```
> 
> Curious to see if this works!

I'll try that.
Thanks for the tips

ewels commented:
@ggabernet - the error message that you got was still kind of crappy: bad string interpolation and it didn't tell you what was wrong. I've just refactored how all Nextflow commands are called into a rewritten function that now gives a much nicer log message. Hopefully I haven't broken everything.

Could you give it a whirl again please?

ewels commented:
Ok, updated so that you now specify `--fix <test-name>` for each test that you want it to try to fix. When running without `--fix` you get a nice tip at the end of the log with a copy + paste command that will run for each type of test that _could_ do something:

![image](https://user-images.githubusercontent.com/465550/107686737-df7f5600-6ca5-11eb-81b1-9b9378ed42c6.png)

But this is hopefully a little more explicit and certainly a lot easier to customise.

ggabernet commented:
oh now it does not fail any more even if my nextflow is broken:
```
$ nextflow
ERROR: Cannot find Java or it's a wrong version -- please make sure that Java 8 or later is installed
NOTE: Nextflow is trying to use the Java VM defined by the following environment variables:
 JAVA_CMD: /Library/Java/JavaVirtualMachines/jdk-15.0.2.jdk/Contents/Home/bin/java
 JAVA_HOME: /Library/Java/JavaVirtualMachines/jdk-15.0.2.jdk/Contents/Home
 ```
 
 ```
 $ nf-core lint .

                                          ,--./,-.
          ___     __   __   __   ___     /,-._.--~\
    |\ | |__  __ /  ` /  \ |__) |__         }  {
    | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                          `._,._,'

    nf-core/tools version 1.13.dev0



INFO     Testing pipeline: .                                             __init__.py:194
╭──────────────────────────────────────────────────────────────────────────────────────╮
│ [?] 2 Tests Ignored                                                                  │
├──────────────────────────────────────────────────────────────────────────────────────┤
│ files_unchanged: File does not exist: assets/nf-core-test_logo.png                   │
│ files_unchanged: File does not exist: conf/charliecloud.config                       │
╰──────────────────────────────────────────────────────────────────────────────────────╯
╭──────────────────────────────────────────────────────────────────────────────────────╮
│ [!] 40 Test Warnings                                                                 │
├──────────────────────────────────────────────────────────────────────────────────────┤
│ actions_awsfulltest: .github/workflows/awsfulltest.yml should test full datasets, n… │
│ conda_env_yaml: Conda dep outdated: conda-forge::python=3.6.11, 3.9.1 available      │
│ conda_env_yaml: Conda dep outdated: conda-forge::markdown=3.1.1, 3.3.3 available     │
│ conda_env_yaml: Conda dep outdated: conda-forge::pymdown-extensions=6.0, 8.1.1 avai… │
│ conda_env_yaml: Conda dep outdated: conda-forge::pygments=2.5.2, 2.7.4 available     │
│ conda_env_yaml: Conda dep outdated: vsearch=2.11.1, 2.15.2 available                 │
│ conda_env_yaml: Conda dep outdated: bioconda::fastqc=0.11.8, 0.11.9 available        │
│ conda_env_yaml: Conda dep outdated: matplotlib=3.0.3, 3.3.4 available                │
│ conda_env_yaml: Conda dep outdated: blast=2.7.1, 2.10.1 available                    │
│ conda_env_yaml: Conda dep outdated: igblast=1.10.0, 1.15.0 available                 │
│ conda_env_yaml: Conda dep outdated: tbl2asn=25.6, 25.7 available                     │
│ conda_env_yaml: Conda dep outdated: airr=1.2.1, 1.3.1 available                      │
│ conda_env_yaml: Conda dep outdated: presto=0.5.10, 0.6.2 available                   │
│ conda_env_yaml: Conda dep outdated: changeo=0.4.5, 1.0.1 available                   │
│ conda_env_yaml: Conda dep outdated: conda-forge::ghostscript=9.22, 9.53.3 available  │
│ conda_env_yaml: Conda dep outdated: conda-forge::pandas=0.24.2, 1.2.2 available      │
│ conda_env_yaml: Conda dep outdated: conda-forge::seaborn=0.9.0, 0.11.1 available     │
│ conda_env_yaml: Conda dep outdated: conda-forge::r-base=3.5.1, 4.0.3 available       │
│ conda_env_yaml: Conda dep outdated: conda-forge::r-alakazam=0.2.11, 1.0.2 available  │
│ conda_env_yaml: Conda dep outdated: bioconda::r-shazam=0.1.11, 0.2.3 available       │
│ conda_env_yaml: Conda dep outdated: bioconda::r-tigger=0.3.1, 0.4.0 available        │
│ conda_env_yaml: Conda dep outdated: conda-forge::r-dplyr=0.8.3, 1.0.4 available      │
│ conda_env_yaml: Conda dep outdated: conda-forge::r-data.table=1.12.2, 1.13.6 availa… │
│ conda_env_yaml: Conda dep outdated: conda-forge::r-igraph=1.2.4.1, 1.2.6 available   │
│ conda_env_yaml: Conda dep outdated: conda-forge::r-gplots=3.0.1.1, 3.1.1 available   │
│ conda_env_yaml: Conda dep outdated: conda-forge::r-circlize=0.4.6, 0.4.12 available  │
│ conda_env_yaml: Conda dep outdated: conda-forge::r-gtools=3.8.1, 3.8.2 available     │
│ conda_env_yaml: Conda dep outdated: conda-forge::r-ggplot2=3.2.0, 3.3.3 available    │
│ conda_env_yaml: Conda dep outdated: conda-forge::r-svglite=1.2.2, 1.2.3 available    │
│ conda_env_yaml: Conda dep outdated: conda-forge::r-optparse=1.6.2, 1.6.6 available   │
│ conda_env_yaml: Conda dep outdated: r-airr=1.2.0, 1.3.0 available                    │
│ pipeline_todos: TODO string in README.md: If applicable, make list of people who ha… │
│ pipeline_todos: TODO string in main.nf: Add in log files from your new processes fo… │
│ pipeline_todos: TODO string in main.nf: Specify which MultiQC modules to use with -… │
│ pipeline_todos: TODO string in usage.md: Add documentation about anything specific … │
│ pipeline_todos: TODO string in awsfulltest.yml: You can customise AWS full pipeline… │
│ pipeline_todos: TODO string in awstest.yml: You can customise CI pipeline run tests… │
│ pipeline_todos: TODO string in ci.yml: You can customise CI pipeline run tests as r… │
│ pipeline_todos: TODO string in test_full.config: Specify the paths to your full tes… │
│ pipeline_todos: TODO string in test_full.config: Give any required params for the t… │
╰──────────────────────────────────────────────────────────────────────────────────────╯
╭──────────────────────────────────────────────────────────────────────────────────────╮
│ [✗] 9 Tests Failed                                                                   │
├──────────────────────────────────────────────────────────────────────────────────────┤
│ files_unchanged: '.github/CONTRIBUTING.md' does not match the template               │
│ files_unchanged: '.github/PULL_REQUEST_TEMPLATE.md' does not match the template      │
│ files_unchanged: '.github/ISSUE_TEMPLATE/bug_report.md' does not match the template  │
│ files_unchanged: '.github/ISSUE_TEMPLATE/feature_request.md' does not match the tem… │
│ files_unchanged: '.github/workflows/linting.yml' does not match the template         │
│ files_unchanged: 'assets/email_template.html' does not match the template            │
│ files_unchanged: 'assets/email_template.txt' does not match the template             │
│ files_unchanged: 'docs/images/nf-core-bcellmagic_logo.png' does not match the templ… │
│ files_unchanged: 'assets/multiqc_config.yaml' does not match the template            │
╰──────────────────────────────────────────────────────────────────────────────────────╯
╭───────────────────────╮
│ LINT RESULTS SUMMARY  │
├───────────────────────┤
│ [✔] 150 Tests Passed  │
│ [?]   2 Tests Ignored │
│ [!]  40 Test Warnings │
│ [✗]   9 Tests Failed  │
╰───────────────────────╯
Tip: Running with '--fix' can automatically resolve some lint failures.
```

ewels commented:
> oh now it does not fail any more even if my nextflow is broken:

This will be because `nf-core` caches the results of the `nextflow config` command as it's pretty slow. So if you got it to work before and haven't changed the the config file since it will skip the call. Try changing some whitespace or something and you should see the error.

ewels commented:
Also, I don't think that you've pulled in the latest changes, as this log message should be different:

> ```
> Tip: Running with '--fix' can automatically resolve some lint failures.
> ```

ggabernet commented:
Ah right that was it:
```
$ nf-core lint .

                                          ,--./,-.
          ___     __   __   __   ___     /,-._.--~\
    |\ | |__  __ /  ` /  \ |__) |__         }  {
    | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                          `._,._,'

    nf-core/tools version 1.13.dev0



CRITICAL Command 'nextflow config -flat .' returned non-zero error code  __main__.py:325
         '1':
         > [31mERROR: Cannot find Java or it's a wrong version -- please
         make sure that Java 8 or later is installed(B[m
         [33mNOTE: Nextflow is trying to use the Java VM defined by the
         following environment variables:
          JAVA_CMD: /Library/Java/JavaVirtualMachines/jdk-15.0.2.jdk/Con
         tents/Home/bin/java
          JAVA_HOME:
         /Library/Java/JavaVirtualMachines/jdk-15.0.2.jdk/Contents/Home
         (B[m
```

ggabernet commented on nf_core/list.py:
```diff
@@ -360,25 +342,12 @@ def get_local_nf_workflow_details(self):
-                        m = re.search(pattern, nfinfo_raw)
-                        if m:
-                            setattr(self, key, m.group(1))
+                nfinfo_raw = str(nf_core.utils.nextflow_cmd(f"nextflow indo -d {self.full_name}"))
```
```suggestion
                nfinfo_raw = str(nf_core.utils.nextflow_cmd(f"nextflow info -d {self.full_name}"))
```

ewels commented:
Ack, doubled-up ASCII colour codes there showing up as `[31m`. Ah well, at least it's sort of readable now 😅 

Thanks for testing!
